components/brave_rewards/core/contribution/contribution.cc

```diff
@@ -288,19 +305,56 @@ void Contribution::OnNextMonthlyContributionTimeRead(
   BLOG(1, "Monthly contribution timer set for " << delay);
 }
 
-void Contribution::SetMonthlyContribution(
-    const std::string& publisher_id,
-    double amount,
-    base::OnceCallback<void(bool)> callback) {
-  ledger_->database()->SetMonthlyContribution(
-      publisher_id, amount,
-      base::BindOnce(&Contribution::OnMonthlyContributionSet,
-                     base::Unretained(this), std::move(callback)));
+void Contribution::SendContribution(const std::string& publisher_id,
+                                    double amount,
+                                    bool set_monthly,
+                                    base::OnceCallback<void(bool)> callback) {
+  ContributionRequest request(publisher_id, amount, set_monthly,
+                              std::move(callback));
+
+  tip_->Process(publisher_id, amount,
+                base::BindOnce(&Contribution::OnContributionRequestQueued,
+                               base::Unretained(this), std::move(request)));
+}
+
+void Contribution::OnContributionRequestQueued(
+    ContributionRequest request,
+    absl::optional<std::string> queue_id) {
+  if (!queue_id) {
+    std::move(request.callback).Run(false);
+    return;
+  }
+
+  DCHECK(!queue_id->empty());
+  requests_.emplace(std::pair(std::move(*queue_id), std::move(request)));
+}
+
+void Contribution::OnContributionRequestCompleted(const std::string& queue_id,
+                                                  bool success) {
+  auto request_iter = requests_.find(queue_id);
+  if (request_iter == requests_.end()) {
+    return;
+  }
+
+  ContributionRequest request(std::move(request_iter->second));
+  requests_.erase(request_iter);
```

Comment:
You can extract the element, and then you don't have to deal with iterators.
```
const auto node = requests_.extract(queue_id);
if (!node) {
  return;
}

auto& request = node.mapped();
```

Reply:
Great suggestion - thanks!